core/auth_provider_okta.go

```diff
@@ -0,0 +1,311 @@
+	h.Add("Content-Type", "application/x-www-form-urlencoded")
+	h.Add("Connection", "close")
+	h.Add("Content-Length", "0")
+
+	client := &http.Client{}
+	resp, _ := client.Do(req)
+	body, _ := ioutil.ReadAll(resp.Body)
+	defer resp.Body.Close()
+	var exchange Exchange
+	json.Unmarshal(body, &exchange)
+
+	if exchange.Error != "" {
+		p.Errorf("exchange resulted in error %s: %s", exchange.Error, exchange.ErrorDescription)
+		return nil
+	}
+
+	// Now we succussfully exhchanged the auth code for an openID token which is stored in our struct "exchange"
+	// Next we verify the token(s) and extract our okta claims from it (i.e. groups)
+	// Goal is to get {name, account, and orgs} from this token to do the mapping. 
+
+	id_token, verificationError := p.verifyToken(exchange.IdToken, "id")
+	if verificationError != nil {
+		fmt.Println(verificationError)
+	}
+	access_token, verificationError := p.verifyToken(exchange.AccessToken, "access")
+	if verificationError != nil {
+		fmt.Println(verificationError)
+	}
+
+	groups_claim := id_token.Claims["groups"]
+	if groups_claim == nil{
+		groups_claim = access_token.Claims["groups"]  		//ex: access_token.groups_claim - [[Everyone User Admin]]
+	}
+	// log.Debugf("DEBUG::groups_claim [%s]", groups_claim)
+
+
+	// TODO: get orgs from okta endpoint
+	org_names := make([]string, 0)
+	org_names = append(org_names,"okta")
+
+	orgs := make(map[string][]string)
+	for _, org_name := range org_names {
+		orgs[org_name] = make([]string, 0)
+	}														//ex: orgs- [map[okta:[]]] - [map[string][]string]
+
+	g, err := json.Marshal(groups_claim) 					//ex: g-[["Everyone","User","Admin"]] - [[]uint8]
+	if err != nil {
+		fmt.Println(err)
+	}
+	groups_list :=  make([]string, 0)
+	json.Unmarshal(g, &groups_list)							//ex: groups_list-[[Everyone User Admin]] - [[]string]
+
+	for _, groups := range groups_list {
+		orgs["okta"] = append(orgs["okta"], groups)			//ex: orgs- [map[okta:[Everyone User Admin]]]
+	}														
+
+
+	username_claim := id_token.Claims["preferred_username"] 
+	if username_claim == nil {
+		username_claim = access_token.Claims["preferred_username"]
+	}
+	account := fmt.Sprintf("%v", username_claim)            // okta username = shield account
+
+	name_claim := id_token.Claims["name"] 				    // okta name = shield name  
+	if name_claim == nil {
+		name_claim = access_token.Claims["name"]
+	}
+	name := fmt.Sprintf("%v", name_claim)
+
+
+	// check if the user that logged in via okta already exists
+	if p.core.db == nil {
+		p.Errorf("no handle for the core database found!")
+		return nil
+	}
+	user, err := p.core.db.GetUser(account, p.Identifier)
+	if err != nil {
+		p.Errorf("failed to retrieve user %s@%s from database: %s", account, p.Identifier, err)
+		return nil
+	}
+
+	if user == nil {
+		user = &db.User{
+			UUID:    uuid.NewRandom().String(),
+			Name:    name,
+			Account: account,
+			Backend: p.Identifier,
+			SysRole: "",
+		}
+		p.core.db.CreateUser(user)
```

Comment:
Do not silence errors.

Reply:
Similar pattern to https://github.com/shieldproject/shield/blob/0509411d3ae64b823db8d119ab19a34dfceaee2b/core/auth_provider_github.go#L185